### DISCUSSION TITLE
How to insert and save high resolution image to PDF

### DISCUSSION THREAD

**Discussion by coderqxr** [pymupdf/PyMuPDF#1004] [Looking for help]
## Describe the bug (mandatory)

My PNG file is 600dpi, which is changed to 72 after storing as PDF.
How can I keep the original 600dpi unchanged when storing PDF?

## To Reproduce (mandatory)

    doc = fitz.open ()  
    image = fitz.Pixmap (filename)
    image.setResolution (600, 600)  
    page = doc.new_ page(0, image.width , image.height )  
    page.insert_ image((0, 0, image.width , image.height ), pixmap=image)  
    newFileName = filename.replace ('.png', '.pdf')  
    doc.save (newFileName)  
    doc.close ()

## Your configuration (mandatory)

- Python 3.8

- PyMuPDF 1.18.9

**Comment by JorjMcKie** [comment]
Try not to use pixmap for insertion, but directly use the file: ``page.insert_image(rect, filename=filename)``.

**Comment by coderqxr** [comment]
It still doesn't work

My PNG image was originally 96dpi. It is effective to use pixmap to change it to 600dpi and then save the PNG with overlay. However, when it is saved as PDF, it becomes 72dpi

    doc = fitz.open()
    image = fitz.Pixmap(filename)  # open pic as document
    image.setResolution(600, 600)
    image.writePNG(filename)
    page = doc.new_page(0, image.width, image.height)
    # page.insert_image((0, 0, image.width, image.height), pixmap=image)
    page.insert_image((0, 0, image.width, image.height), filename=filename)
    newFileName = file.replace('.png', '.pdf')
    doc.save(output_path + "\\" + newFileName)
    doc.close()

**Comment by JorjMcKie** [comment]
Hm. Is your PNG transparent? If yes, then internally still a pixmap is generated to compute the separately required transparency mask image.
If not, you could try and convert the PNG to a JPG (using PIL), simultaneously setting the dpi and insert the JPG binary image using the `stream` argument.

**Comment by JorjMcKie** [comment]
Meanwhile, I will look in the code to see if there is a way to enforce the dpi.

**Comment by coderqxr** [comment]
Oh, yes, my PNG has a transparent layer.

So, how to save PNG with transparent layer as PDF and keep DPI unchanged?

**Comment by coderqxr** [comment]
I tried another way to insert the image, and the DPI was preserved, but the transparent layer was lost

    doc = fitz.open()
    imgdoc = fitz.open(filename)
    pdfbytes = imgdoc.convertToPDF()
    imgpdf = fitz.open("pdf", pdfbytes)
    doc.insertPDF(imgpdf)
    newFileName = file.replace('.png', '.pdf')
    doc.save(output_path + "\\" + newFileName)
    doc.close()

So, is there any way to modify the DPI and keep the transparent layer at the same time?

**Comment by JorjMcKie** [comment]
I made a few code adjustments. What is your full environemnt (print(fitz.__doc__) output)?
I will generate a wheel to test for you.

**Comment by JorjMcKie** [comment]
ah, looks like Windows ...

**Comment by JorjMcKie** [comment]
[PyMuPDF-1.18.11-cp38-cp38-win_amd64.whl.zip](https://github.com/pymupdf/PyMuPDF/files/6265985/PyMuPDF-1.18.11-cp38-cp38-win_amd64.whl.zip)

**Comment by coderqxr** [comment]
Yes, it's Windows.
PyMuPDF-1.18.11still doesn't work.
My full environemnt:

    PyMuPDF 1.18.11: Python bindings for the MuPDF 1.18.0 library.
    Version date: 2021-04-04 07:00:24.
    Built for Python 3.8 on win32 (64-bit).

My IDE is PyCharm, and Python interpreter is configured as:
![image](https://user-images.githubusercontent.com/12953038/113800541-ba7d0000-9789-11eb-93f5-831d06e93df3.png)

**Comment by JorjMcKie** [comment]
I am still trying a few things. The current status is this:
* Non-transparent images are inserted with their original dpi - whether filename, stream or pixmap is used
* Images with an alpha channel receive a dpi of (96, 96), no matter what I try to enforce the original value.

The only option left is the following approach:
* If an image has transparency, split it up in two images:
    1. Image with the color components only ("image1")
    2. Image with the transparency information ("image2")
* Use `page.insert_image(rect, stream=image1, mask=image2)`

This would hopefully prevent / circumvent whatever is happening otherwise.
The image parameters above must both be valid binary image formats (as Python `bytes` objects). I am not sure, which format to choose for image2. Definitely not PNG, so pillow would have to be used to create an appropriate `PIL.Image` from the `samples` contained in image2.

**Comment by coderqxr** [comment]
OK, I'll try later

**Comment by coderqxr** [comment]
I failed.   
Using transparent layer as mask will cover the original image of other channels.  
I also tried to read the PNG file by PIL and convert them to bytes. When I save it as PDF, the 600 DPI will still be changed to 72 DPI.  
Is this a bug, and will it be fixed in later versions? If the later version will fix this problem, I think I can wait a moment~  
![image](https://user-images.githubusercontent.com/12953038/114036919-0d060b80-98b3-11eb-998a-3bdd409e1eb9.png)

**Comment by JorjMcKie** [comment]
It definitely happens inside MuPDF, not PyMuPDF.
I am not aware of any plans at MuPDF doing something about it - or even whether they are aware of this feature gap.
I don't call it a bug, because there is no assertion anywhere that it is supposed to work ...

**Comment by coderqxr** [comment]
OK, thank you so much. I will continue to pay attention to this issue.
If you know about the solution or improvement of this problem, please let me know in time.
thanks~   ^_^

**Comment by JorjMcKie** [comment]
Of course - thanks for your patience!

**Comment by vigneshk01** [comment]
Hello, 
Do we have any updates or fix for this issue.
Thanks in advance.

**Comment by JorjMcKie** [comment]
What is your platform?
I just generated pre-version 1.19.3 Linux to [here](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse).
Can do that for OSX as well, if required, or upload your Windows one.
I trust that the issue is being solved with this version.

**Reply by vigneshk01** [reply]
Great, thanks for the update. 

I am using Windows. Below are the details:
`PyMuPDF 1.19.2: Python bindings for the MuPDF 1.19.0 library.`
`Version date: 2021-11-20 00:00:01.`
`Built for Python 3.9 on win32 (64-bit).`

**Comment by vigneshk01** [comment]
![image](https://user-images.githubusercontent.com/33518968/143691223-9eebccb4-c11c-4101-a8a6-4ba27b0a0199.png)

In the mean time, I tried the above given code. 
It generates the transparent image on my pdf, but appears greyed out, 
Any suggestions on how to bring the color on the transparent image?.

Thanks.

**Reply by JorjMcKie** [reply]
Then insert your image (as file or stream) as is - no prior conversion to pixmap.
No need to specify alpha, or whatever.

**Comment by JorjMcKie** [comment]
[PyMuPDF-1.19.3-cp39-cp39-win_amd64.zip](https://github.com/pymupdf/PyMuPDF/files/7612322/PyMuPDF-1.19.3-cp39-cp39-win_amd64.zip)
This is the complete upload

**Reply by vigneshk01** [reply]
Thanks a lot for sharing.

Also as per one of your suggestions, I tried the below by replacing the stream argument with the actual stream of the image, I could see the color on the transparent image.
![image](https://user-images.githubusercontent.com/33518968/143898265-099410cb-3c80-4771-abe8-6b72f733a344.png)
